From ef1f43b3e39da70ba3a65053cd03470c5b3486df Mon Sep 17 00:00:00 2001
From: Phil Staub <Phil.Staub@windriver.com>
Date: Thu, 10 Mar 2011 11:38:38 -0800
Subject: [PATCH 08/13] Cavium: Reduce jitter when reading PTP clock

Source: Cavium sdk_2.0.0_updates_p4.tgz

In cvm_oct_ptp_to_ktime(), ktimebase is read with ktime_get_real(),
and ptpbase is read with octeon_read_ptp_csr(). Arrange for
ktime_get_real to be in icache to minimize execution time
variability. This is done because ktimebase and ptpbase must be read
as close to the exact same time as possible.

Signed-off-by: Phil Staub <Phil.Staub@windriver.com>
Signed-off-by: ltian <le.tian@windriver.com>
---
 drivers/net/octeon/ethernet-rx.c |    7 +++++++
 1 files changed, 7 insertions(+), 0 deletions(-)

diff --git a/drivers/net/octeon/ethernet-rx.c b/drivers/net/octeon/ethernet-rx.c
index 825d8d3..6becdc4 100644
--- a/drivers/net/octeon/ethernet-rx.c
+++ b/drivers/net/octeon/ethernet-rx.c
@@ -239,6 +239,13 @@ static ktime_t cvm_oct_ptp_to_ktime(uint64_t ptptime)
 	unsigned long flags;
 
 	local_irq_save(flags);
+	/* Fill the icache with the code */
+	ktime_get_real();
+	/* Flush all pending operations */
+	CVMX_SYNC;
+	/* Read the time and PTP clock as close together as possible. It is
+	   important that this sequence take the same amount of time to
+	   reduce jitter */
 	ktimebase = ktime_get_real();
 	ptpbase = octeon_read_ptp_csr(CVMX_MIO_PTP_CLOCK_HI);
 	local_irq_restore(flags);
-- 
1.7.0.4

